### PR TITLE
DoubleClick: add double-click

### PR DESCRIPTION
**Pull request by RuffaloLavoisier**
~~TEST sample code~~
Just use 
```
hal->btnIsDoubleClick(BUTTON_N)
```

### CODE REVIEW

File: src/hal/buttons.cpp [hunk 1]
```diff
@@ -81,6 +92,10 @@ void OswHal::checkButtons(void) {
 bool OswHal::btnHasGoneDown(Button btn) {
     return _btnGoneDown[btn];
 }
+bool OswHal::btnDoublePress(Button btn) {
+    Serial.println(_btnDoublePress[btn]);
```
Comment (uvwxy) [inline src/hal/buttons.cpp:96]: should be removed, or wrapped with a debug define

File: include/osw_hal.h [hunk 2]
```diff
@@ -66,6 +66,7 @@ class OswHal {
 
     // Buttons (Engine-Style)
     bool btnHasGoneDown(Button btn);
+    bool btnDoublePress(Button btn);
```
Comment (uvwxy) [inline include/osw_hal.h:69]: maybe better: btnIsDoubleClick ?

the PR talks about double-click, the code about press.

File: include/osw_hal.h [hunk 3]
```diff
@@ -221,6 +222,12 @@ class OswHal {
     bool _btnSuppressUntilUpAgain[NUM_BUTTONS];
     bool _btnGoneDown[NUM_BUTTONS];
     unsigned long _btnIsDownMillis[NUM_BUTTONS];
+
+    bool double_timeOut[NUM_BUTTONS];
```
Comment (uvwxy) [inline include/osw_hal.h:226]: _btnDoubleClickTimeout

File: include/osw_hal.h [hunk 4]
```diff
@@ -221,6 +222,12 @@ class OswHal {
...
+    unsigned long _btnDoubleMillis[NUM_BUTTONS];
```
Comment (uvwxy) [inline include/osw_hal.h:227]: _btnDoubleClickMillis

File: include/osw_hal.h [hunk 5]
```diff
@@ -221,6 +222,12 @@ class OswHal {
...
+    bool _btnDoublePress[NUM_BUTTONS];
```
Comment (uvwxy) [inline include/osw_hal.h:228]: _btnDoubleClick

File: include/osw_hal.h [hunk 6]
```diff
@@ -221,6 +222,12 @@ class OswHal {
...
+    uint8_t _detectDoubleCount[NUM_BUTTONS];
```
Comment (uvwxy) [inline include/osw_hal.h:229]: _btnDetectDoubleClickCount ?

File: src/apps/watchfaces/OswAppWatchface.cpp [hunk 7]
```diff
@@ -136,9 +136,12 @@ void OswAppWatchface::loop() {
     if (hal->btnHasGoneDown(BUTTON_3)) {
         hal->increaseBrightness(25);
     }
-    if (hal->btnHasGoneDown(BUTTON_2)) {
+     if (hal->btnDoublePress(BUTTON_2)) {
```
Comment (uvwxy) [inline src/apps/watchfaces/OswAppWatchface.cpp:139]: is this for testing or do we want to change this?

File: src/hal/buttons.cpp [hunk 8]
```diff
@@ -81,6 +92,10 @@ void OswHal::checkButtons(void) {
 bool OswHal::btnHasGoneDown(Button btn) {
     return _btnGoneDown[btn];
 }
+bool OswHal::btnDoublePress(Button btn) {
```
Comment (uvwxy) [inline src/hal/buttons.cpp:95]: btnIsDoubleClick

**Review by uvwxy — CHANGES_REQUESTED** [review]
Thank you for your contribution, I think we could organise this a little regarding the existing naming of button events/variables :)

File: src/apps/watchfaces/OswAppWatchface.cpp (hunk 7 above)
Comment (RuffaloLavoisier) [reply src/apps/watchfaces/OswAppWatchface.cpp:139]: Just test :) plz check it next weekend !

**Comment by RuffaloLavoisier** [comment]
Why not working Github......

File: src/apps/watchfaces/OswAppWatchface.cpp [hunk 9]
```diff
@@ -139,6 +142,9 @@ void OswAppWatchface::loop() {
     if (hal->btnHasGoneDown(BUTTON_2)) {
         hal->decreaseBrightness(25);
     }
+    if (hal->btnIsDoubleClick(BUTTON_2)) {
+        OswAppWatchface::settingDisplayMainWatchface("0");
```
Comment (RuffaloLavoisier) [inline src/apps/watchfaces/OswAppWatchface.cpp:146]: If this looks good, I want to apply it to all watchfaces. Also, there must be a way to get the current watchface index, can anyone tell me? It's a hard-coding method right now.

File: src/hal/buttons.cpp [hunk 10]
```diff
@@ -52,10 +54,18 @@ void OswHal::checkButtons(void) {
...
-
+        _btnDoubleClick[i] = false;
         // store the time stamp since the button went down
         if (_btnGoneDown[i]) {
             _btnIsDownMillis[i] = millis();
+            if(_btnDetectDoubleClickCount[i]==0){
+                _btnDoubleClickMillis[i] = millis();
+            }
+            _btnDetectDoubleClickCount[i] = _btnDetectDoubleClickCount[i] + 1;
+            _btnDoubleClick[i] = 2 == _btnDetectDoubleClickCount[i] % 3;
```
Comment (RuffaloLavoisier) [inline src/hal/buttons.cpp:65]: Hmmm..sh*t

File: src/apps/watchfaces/OswAppWatchface.cpp (hunk 9 above)
Comment (simonmicro) [reply src/apps/watchfaces/OswAppWatchface.cpp:146]: https://en.cppreference.com/w/cpp/string/byte/strtol

https://en.cppreference.com/w/cpp/string/basic_string/to_string

File: src/apps/watchfaces/OswAppWatchface.cpp [hunk 11]
```diff
@@ -55,7 +55,11 @@ void OswAppWatchface::drawStepHistory(OswUI* ui, uint8_t x, uint8_t y, uint8_t w
     }
 }
 #endif
-
+void OswAppWatchface::settingDisplayMainWatchface(String watchFaceIndex){
```
Comment (simonmicro) [inline src/apps/watchfaces/OswAppWatchface.cpp:58]: As there is noone else using this -> could you merge that into the double-click handling of `handleButtonDefaults()`?

File: src/hal/buttons.cpp [hunk 12]
```diff
@@ -52,10 +54,18 @@ void OswHal::checkButtons(void) {
...
         // store the time stamp since the button went down
         if (_btnGoneDown[i]) {
             _btnIsDownMillis[i] = millis();
+            if(_btnDetectDoubleClickCount[i]==0){
+                _btnDoubleClickMillis[i] = millis();
+            }
+            _btnDetectDoubleClickCount[i] = _btnDetectDoubleClickCount[i] + 1;
+            _btnDoubleClick[i] = 2 == _btnDetectDoubleClickCount[i];
+        }
+        if(millis() - _btnDoubleClickMillis[i] > 1000) {
```
Comment (simonmicro) [inline src/hal/buttons.cpp:67]: 1 second "cooldown"? This seems a little bit much!

File: src/apps/watchfaces/OswAppWatchface.cpp [hunk 13]
```diff
@@ -140,6 +144,8 @@ void OswAppWatchface::handleButtonDefaults() {
         OswHal::getInstance()->increaseBrightness(25);
     if (OswHal::getInstance()->btnHasGoneDown(BUTTON_2))
         OswHal::getInstance()->decreaseBrightness(25);
+    if (OswHal::getInstance()->btnIsDoubleClick(BUTTON_2))
```
Comment (simonmicro) [inline src/apps/watchfaces/OswAppWatchface.cpp:147]: I think there will be problems with users trying to decrease their screen brightness and accidentally overwriting their favorite (default) watchface.

**Review by simonmicro — CHANGES_REQUESTED** [review]

File: src/apps/watchfaces/OswAppWatchface.cpp (hunk 13 above)
Comment (simonmicro) [reply src/apps/watchfaces/OswAppWatchface.cpp:147]: I think there is the problem, that users tend to really spam the brightness control buttons whenever they _quickly_ want to change it!

Ideas:

1. Try to use long press instead for the default watchface setting. There is still no visual indicator, about what was done, but still better than colliding controls!
2. #266 wanted to implement that anyways, right? Sou the watchface could add an additional multi-button mode for that. Whenever the user switches modes, a little indicator icon will show up temporarily to communicate which mode the buttons are an - this also does not collide with the brightness control...

File: src/hal/buttons.cpp [hunk 14]
```diff
@@ -52,10 +54,18 @@ void OswHal::checkButtons(void) {
...
         // store the time stamp since the button went down
         if (_btnGoneDown[i]) {
             _btnIsDownMillis[i] = millis();
+            if(_btnDetectDoubleClickCount[i]==0){
+                _btnDoubleClickMillis[i] = millis();
+            }
+            _btnDetectDoubleClickCount[i] = _btnDetectDoubleClickCount[i] + 1;
+            _btnDoubleClick[i] = 2 == _btnDetectDoubleClickCount[i];
+        }
+        if(millis() - _btnDoubleClickMillis[i] > 500) {
```
Comment (simonmicro) [inline src/hal/buttons.cpp:67]: So, half a second now? What happened to your argument of wearing buttons?

**Review by simonmicro — APPROVED** [review]

**Pull request merged**